src/modules/frost/keygen_impl.h

```diff
@@ -0,0 +1,161 @@
+    secp256k1_sha256_finalize(&sha, buf);
+    secp256k1_scalar_set_b32(indexhash, buf, NULL);
+    /* The x-coordinate must not be zero (see
+     * draft-irtf-cfrg-frost-08#section-4.2.2) */
+    if (secp256k1_scalar_is_zero(indexhash)) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static const unsigned char secp256k1_frost_share_magic[4] = { 0xa1, 0x6a, 0x42, 0x03 };
+
+static void secp256k1_frost_share_save(secp256k1_frost_share* share, secp256k1_scalar *s) {
+    memcpy(&share->data[0], secp256k1_frost_share_magic, 4);
+    secp256k1_scalar_get_b32(&share->data[4], s);
+}
+
+static int secp256k1_frost_share_load(const secp256k1_context* ctx, secp256k1_scalar *s, const secp256k1_frost_share* share) {
+    int overflow;
+
+    ARG_CHECK(secp256k1_memcmp_var(&share->data[0], secp256k1_frost_share_magic, 4) == 0);
+    secp256k1_scalar_set_b32(s, &share->data[4], &overflow);
+    /* Parsed shares cannot overflow */
+    VERIFY_CHECK(!overflow);
+    return 1;
+}
+
+int secp256k1_frost_share_serialize(const secp256k1_context* ctx, unsigned char *out32, const secp256k1_frost_share* share) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(out32 != NULL);
+    ARG_CHECK(share != NULL);
+    memcpy(out32, &share->data[4], 32);
+    return 1;
+}
+
+int secp256k1_frost_share_parse(const secp256k1_context* ctx, secp256k1_frost_share* share, const unsigned char *in32) {
+    secp256k1_scalar tmp;
+    int overflow;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(share != NULL);
+    ARG_CHECK(in32 != NULL);
+
+    secp256k1_scalar_set_b32(&tmp, in32, &overflow);
+    if (overflow) {
+        return 0;
+    }
+    secp256k1_frost_share_save(share, &tmp);
+    return 1;
+}
+
+int secp256k1_frost_share_gen(const secp256k1_context *ctx, secp256k1_pubkey *vss_commitment, secp256k1_frost_share *share, const unsigned char *session_id, const secp256k1_keypair *keypair, const secp256k1_xonly_pubkey *pk, size_t threshold) {
+    secp256k1_sha256 sha;
+    secp256k1_scalar idx;
+    secp256k1_scalar sk;
+    secp256k1_scalar share_i;
+    secp256k1_ge ge_tmp;
+    unsigned char buf[32];
+    unsigned char rngseed[32];
+    secp256k1_scalar rand[2];
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(share != NULL);
+    memset(share, 0, sizeof(*share));
+    ARG_CHECK(session_id != NULL);
+    ARG_CHECK(keypair != NULL);
+    ARG_CHECK(pk != NULL);
+    ARG_CHECK(threshold > 1);
+
+    if (!secp256k1_keypair_load(ctx, &sk, &ge_tmp, keypair)) {
+        return 0;
+    }
+    /* The first coefficient is the secret key, and thus the first commitment
+     * is the public key. */
+    if (vss_commitment != NULL) {
+        secp256k1_pubkey_save(&vss_commitment[0], &ge_tmp);
+    }
+    /* Compute seed which commits to threshold and session ID */
+    secp256k1_scalar_get_b32(buf, &sk);
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, session_id, 32);
```

Comment:
So the session_id is used as a seed to generate the coefficients? if so this *must* have ~128 bits of secure randomness, so we should document that this must be uniformly random and not just recommend this.